Add MsgPack implementation in contrib

shssoichiro opened the pull request:
Fixes #149. This is very similar to the existing JSON implementation, since they provide the same type of functionality.

SergioBenitez commented on contrib/src/lib.rs:
```diff
@@ -14,6 +14,7 @@
 //! an asterisk next to the features that are enabled by default:
 //!
 //! * [json*](struct.JSON.html)
+//! * [msgpack*](struct.MsgPack.html)
```
This shouldn't have a `*` on it.

SergioBenitez commented on contrib/src/msgpack/mod.rs:
```diff
@@ -0,0 +1,118 @@
+
+pub use self::rmp_serde::decode::Error as MsgPackError;
+
+/// The MsgPack type: implements `FromData` and `Responder`, allowing you to easily
```
It looks like you just replaced `JSON` with `MsgPack` here, but the wording doesn't work in some cases. Can you customize this for `MsgPack`? In particular, you should likely refer to it as `MessagePack`.

SergioBenitez commented on lib/src/http/content_type.rs:
```diff
@@ -104,6 +104,7 @@ impl ContentType {
         "HTML", HTML, is_html => "text", "html" ; "charset=utf-8",
         "Plain", Plain, is_plain => "text", "plain" ; "charset=utf-8",
         "JSON", JSON, is_json => "application", "json",
+        "MsgPack", MsgPack, is_msgpack => "application", "msgpack",
```
The internet seems to think that the correct Content-Type is `application/x-msgpack`. Why did you choose this?

shssoichiro replied:
The reason I used `application/msgpack` is because [the x- prefix is deprecated](https://tools.ietf.org/html/rfc6648) on MIME types. The most recent/most upvoted answer on [this StackOverflow question](http://stackoverflow.com/questions/6888900/mime-type-for-msgpack#25886130) recommends using `application/msgpack` for this reason. However, I've found `application/x-msgpack` and even `bin/msgpack` used in other places on a Google search, and it seems that msgpack [never registered](https://www.iana.org/assignments/media-types/media-types.xhtml) or [decided on](https://github.com/msgpack/msgpack/issues/194) an official MIME type. So, I'll defer to your judgment on which MIME type we should use here.

SergioBenitez replied:
`application/msgpack` seems like what we should use here. But see my inline comments about what we should accept.

SergioBenitez commented on lib/src/response/content.rs:
```diff
@@ -82,6 +82,7 @@ macro_rules! ctrs {
 ctrs! {
     JSON: "JSON", "application/json",
     XML: "XML", "text/xml",
+    MsgPack: "MsgPack", "application/msgpack",
```
Same thing here.

SergioBenitez replied:
Name is `MessagePack` not "MsgPack`.

SergioBenitez commented on lib/src/response/responder.rs:
```diff
@@ -201,6 +201,16 @@ impl Responder<'static> for String {
 
+/// Returns a response with a fixed-size body
+/// containing the byte vec `self`. Always returns `Ok`.
+impl Responder<'static> for Vec<u8> {
```
Why do we need this? Why can't we just use `Cursor::new(vec)` in the `MsgPack` `Responder`?

shssoichiro replied:
Because the `Responder` wants to return a `Result` with `rocket::Response`, so `Cursor<Vec<u8>>` won't fit that type definition.

SergioBenitez replied:
See my inline comments.

SergioBenitez commented on contrib/src/msgpack/mod.rs:
```diff
@@ -0,0 +1,118 @@
+    type Error = MsgPackError;
+
+    fn from_data(request: &Request, data: Data) -> data::Outcome<Self, Self::Error> {
+        if !request.content_type().map_or(false, |ct| ct.is_msgpack()) {
```
I think the right thing to do here is to match against all of the commonly used MessagePack content types, and accept all of them. Document which are accepted in the top-level docs _and_ here.

SergioBenitez commented on contrib/src/msgpack/mod.rs:
```diff
@@ -0,0 +1,118 @@
+        }
+
+        let mut buf = Vec::new();
+        if let Err(e) = data.open().read_to_end(&mut buf) {
```
This is bad as it opens the application up to DOS attack via an out-of-memory problem. The client can just stream infinite data until the machine runs out of memory. This needs to be limited.

SergioBenitez commented on contrib/src/msgpack/mod.rs:
```diff
@@ -0,0 +1,118 @@
+            error_!("Couldn't read request data: {:?}", e);
+            return Outcome::Failure((Status::BadRequest, e));
+        };
+        match rmp_serde::from_slice(&buf).map(|val| MsgPack(val)) {
```
Will `rmp_serde` not operate on a stream directly?

shssoichiro replied:
It does not appear to have a method for that, no

SergioBenitez commented on contrib/src/msgpack/mod.rs:
```diff
@@ -0,0 +1,118 @@
+impl<T: Serialize> Responder<'static> for MsgPack<T> {
+    fn respond(self) -> response::Result<'static> {
+        rmp_serde::to_vec(&self.0).map(|buf| {
+            content::MsgPack(buf).respond().unwrap()
```
Construct the response here directly so we don't need to `Vec` `Responder` implementation.

SergioBenitez commented on contrib/src/msgpack/mod.rs:
```diff
@@ -0,0 +1,118 @@
+pub use self::rmp_serde::decode::Error as MsgPackError;
+
+/// The `MsgPack` type: implements `FromData` and `Responder`, allowing you to easily
+/// consume and respond with MessagePack.
```
Should say "MessagePack data."

shssoichiro commented:
Pushed the next set of changes. I'm not sure that I did the content type checking in the best possible way, since the `ctrs!` macro doesn't support checking for multiple types, but I needed to leave the MsgPack type in the macro or other parts of the feature would break.

SergioBenitez commented on contrib/src/msgpack/mod.rs:
```diff
@@ -0,0 +1,131 @@
+    fn from_data(request: &Request, data: Data) -> data::Outcome<Self, Self::Error> {
+        // Accepted content types are:
+        // `application/msgpack`, `application/x-msgpack`, and `bin/msgpack`
+        if !request.content_type().map_or(false, |ct|
```
This is really hard to read. Can we split this into another function and reformat it so it's obvious what's happening? Maybe use a `match`? Due to the `UncasedAscii`, I don't know if Rust will allow something like:

```rust
match (ct.ttype, ct.subtype) {
    ("application", "msgpack") | ("application", "x-msgpack") => true,
    ("bin", "msgpack") | ("bin", "x-msgpack") => true,
    _ => false
}
```

But play around with it; maybe something like that will work.

shssoichiro replied:
Looks like `match` needs the type to be exactly the same, so even though `UncasedAscii` implements `PartialEq` for `str`, Rust's typechecker won't allow this.

I did modify it slightly in a way that improves readability, although I don't think it's as nice as this.

SergioBenitez commented on contrib/src/msgpack/mod.rs:
````diff
@@ -0,0 +1,131 @@
+///     MsgPack(user_from_id)
+/// }
+/// ```
+///
````
Kill this empty line here.

SergioBenitez commented on contrib/src/msgpack/mod.rs:
````diff
@@ -0,0 +1,131 @@
+/// fn new_user(user: MsgPack<User>) {
+///     ...
+/// }
+/// ```
````
Add an empty line below this.

SergioBenitez commented on contrib/src/msgpack/mod.rs:
```diff
@@ -0,0 +1,131 @@
+            error_!("Couldn't read request data: {:?}", e);
+            return Outcome::Failure((Status::BadRequest, e));
+        };
+        match rmp_serde::from_slice(&buf).map(|val| MsgPack(val)) {
```
Add an empty line above here.

SergioBenitez commented on contrib/src/msgpack/mod.rs:
```diff
@@ -0,0 +1,132 @@
+    }
+}
+
+// Serializes the wrapped value into MessagePack. Returns a response with Content-Type
```
This should be a real docstring, not a comment. Use three `///`. I've fixed this in the `JSON` docs.

SergioBenitez commented on contrib/src/msgpack/mod.rs:
```diff
@@ -0,0 +1,132 @@
+        rmp_serde::to_vec(&self.0).map(|buf| {
+            Response::build()
+                .sized_body(Cursor::new(buf))
+                .ok::<Status>()
```
Can Rust not infer `Status`?

SergioBenitez commented on contrib/src/msgpack/mod.rs:
```diff
@@ -0,0 +1,132 @@
+// `Err` of `Status::InternalServerError` is returned.
+impl<T: Serialize> Responder<'static> for MsgPack<T> {
+    fn respond(self) -> response::Result<'static> {
+        rmp_serde::to_vec(&self.0).map(|buf| {
```
I think if you use `and_then` instead of `map` and remove the `unwrap`, things might work out okay and you'd be able to get rid of the type annotation. You'd need to move the `map_err` before the `and_then` call for this to work, I believe.

shssoichiro commented:
Fixed that round of comments and added examples. I also needed to add a method to get a response body as a byte Vec, since MessagePack data isn't representable as a UTF-8 String.

SergioBenitez commented on examples/msgpack/src/tests.rs:
```diff
@@ -0,0 +1,105 @@
+use std::collections::HashMap;
+
+#[derive(Deserialize)]
+#[allow(dead_code)]
```
Why allow dead code?...

shssoichiro replied:
This should be just for the id field. The example returns an id field, but that field is never used in the tests.

SergioBenitez commented on examples/msgpack/src/tests.rs:
```diff
@@ -0,0 +1,105 @@
+    // Change the message contents.
+    let req = MockRequest::new(Put, "/message/1")
+        .header(ContentType::MsgPack)
+        .body_binary(rmp_serde::to_vec(&[("contents", "Bye bye, world!")].iter().cloned().collect::<HashMap<&'static str, &'static str>>()).unwrap());
```
This line is way too long.

SergioBenitez commented on examples/msgpack/src/tests.rs:
```diff
@@ -0,0 +1,105 @@
+    let req = MockRequest::new(Get, "/message/1") .header(ContentType::MsgPack);
+    run_test!(&rocket, req, |mut response: Response| {
+        assert_eq!(response.status(), Status::Ok);
+        let body = rmp_serde::from_slice::<Message>(&response.body().unwrap().into_bytes().unwrap()).unwrap();
```
This line is way too long.

SergioBenitez commented on examples/msgpack/src/tests.rs:
```diff
@@ -0,0 +1,105 @@
+    // Add a new message with ID 1.
+    let req = MockRequest::new(Post, "/message/1")
+        .header(ContentType::MsgPack)
+        .body_binary(rmp_serde::to_vec(&[("contents", "Hello, world!")].iter().cloned().collect::<HashMap<&'static str, &'static str>>()).unwrap());
```
Line length. This should probably be split out in a help function or something.

SergioBenitez commented on examples/msgpack/src/tests.rs:
```diff
@@ -0,0 +1,105 @@
+    let req = MockRequest::new(Get, "/message/1") .header(ContentType::MsgPack);
+    run_test!(&rocket, req, |mut response: Response| {
+        assert_eq!(response.status(), Status::Ok);
+        let body = rmp_serde::from_slice::<Message>(&response.body().unwrap().into_bytes().unwrap()).unwrap();
```
Line length.

SergioBenitez commented on lib/src/testing.rs:
````diff
@@ -229,6 +229,28 @@ impl<'r> MockRequest<'r> {
+    ///     .body_binary(&[0x92, 0x2a, 0xaa, 0x74, 0x68, 0x65, 0x20, 0x41, 0x6e, 0x73, 0x77, 0x65, 0x72]);
+    /// ```
+    #[inline]
+    pub fn body_binary<S: AsRef<[u8]>>(mut self, body: S) -> Self {
````
I don't like the idea of having two `body` methods. `str` and `String` are `AsRef<[u8]>`, so we can just use those bounds instead.

SergioBenitez commented on lib/src/response/response.rs:
```diff
@@ -54,7 +54,7 @@ impl<T> Body<T> {
+    /// Attempts to read `self` into a `String` and returns it. If reading or
     /// conversion fails, returns `None`.
     pub fn into_string(self) -> Option<String> {
         let (mut body, mut string) = match self {
```
If you're going to add `into_bytes()`, you should just call `into_bytes().map(String::from_utf8)` here.

SergioBenitez replied:
I've added an `into_bytes` in master.

shssoichiro commented:
The primary use case of Msgpack is sending data from one backend service to another, e.g. from one Rocket app to another using a library such as `reqwest`. However, the setup on the receiving side doesn't change from what's shown in the example. The difference would be that instead of sending data from a browser, it's sent from another app, but it's still just through an HTTP request. So I don't think there's much benefit in us showing that in the example.

I'm about to push a commit with the other changes mentioned.

SergioBenitez commented:
My main point is that the example is a lot more complex than it needs to be. Let's just parse some msgpack and return a message proving that it was parsed correctly and have that be it.

SergioBenitez commented:
Ah, nice! That's much better.

Can you add something that shows deserialization as well? Maybe just an `echo` that return the ID and contents as a string or something?

shssoichiro commented:
How's that?

SergioBenitez commented on examples/msgpack/src/tests.rs:
```diff
@@ -0,0 +1,48 @@
+extern crate rmp_serde;
```
Can we not depend on the crate here? I liked that the msgpack binary was in the test case. Just the binary with a note of what the bytes are representing would be great.

shssoichiro replied:
I don't believe we can decode the response from the `msgpack_get` test, on line 26 of tests.rs, without this crate, and that whole test is rather pointless without that ability.

SergioBenitez replied:
No need to decode. We should know the bytes, right? Just check against the bytes.

SergioBenitez commented on examples/msgpack/src/main.rs:
```diff
@@ -0,0 +1,39 @@
+}
+
+#[post("/", data = "<data>", format = "application/msgpack")]
+fn create(data: MsgPack<Message>) -> Result<(), ()> {
```
This isn't testable. Just return the contents directly (so return type of this function is `String`) so you can test for them later:

```
data.into_inner().contents
```

SergioBenitez commented on examples/msgpack/Cargo.toml:
```diff
@@ -0,0 +1,19 @@
+[dependencies]
+rocket = { path = "../../lib" }
+rocket_codegen = { path = "../../codegen" }
+rmp-serde = "0.12"
```
Let's remove this dep.

shssoichiro commented:
That should be everything.

SergioBenitez commented on contrib/src/msgpack/mod.rs:
```diff
@@ -0,0 +1,134 @@
+/// doesn't specify "application/msgpack" as its first `Content-Type:` header
+/// parameter will not be routed to this handler. By default, Rocket will accept a
+/// Content Type of any of the following for MessagePack data:
+/// `application/msgpack`, `application/x-msgpack`, or `bin/msgpack`.
```
We also accept `bin/x-msgpack`.

SergioBenitez commented on contrib/src/msgpack/mod.rs:
```diff
@@ -0,0 +1,134 @@
+
+/// Maximum size of MessagePack data is 1MB.
+/// TODO: Determine this size from some configuration parameter.
+const MAX_SIZE: u64 = 1048576;
```
Is this a good default limit?

shssoichiro replied:
This is the same that we're using for JSON, and since MessagePack is more efficient, this actually gets us a little more data.

SergioBenitez replied:
I was thinking that maybe MessagePack encoded messages tend to contain more information, but let's go with this for now.

SergioBenitez commented on examples/msgpack/src/main.rs:
```diff
@@ -0,0 +1,38 @@
+
+#[post("/", data = "<data>", format = "application/msgpack")]
+fn create(data: MsgPack<Message>) -> Result<String, ()> {
+    Ok(data.contents.clone())
```
Why clone? `data.into_inner().contents` would be better.

SergioBenitez commented:
Thanks so much for this, again! Merged in d43678c35ea3cbaab9343faf1565638f44f94570.
